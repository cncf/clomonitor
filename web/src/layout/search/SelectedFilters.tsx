import { isEmpty } from 'lodash';
import { Fragment } from 'react';
import { IoMdCloseCircleOutline } from 'react-icons/io';

import { FILTERS } from '../../data';
import { Category, Filter, FilterKind, FiltersSection, Maturity } from '../../types';
import styles from './SelectedFilters.module.css';

interface Props {
  filters: { [key: string]: (string | number)[] };
  onChange: (name: string, value: string, checked: boolean) => void;
}

const SelectedFilters = (props: Props) => {
  if (isEmpty(props.filters)) return null;

  const getFilterName = (type: FilterKind, filter: string | number): string => {
    switch (type) {
      case FilterKind.Category:
        return Category[parseInt(filter as string)];
      case FilterKind.Maturity:
        return Maturity[parseInt(filter as string)];
      case FilterKind.Rating:
        let ratingName: string = '';
        const ratings = FILTERS.find((sec: FiltersSection) => sec.name === type);
        if (ratings) {
          const rating = ratings.filters.find((f: Filter) => f.name === filter);
          if (rating) {
            ratingName = `${rating.label} ${rating.legend}`;
          }
        }
        return ratingName;
    }
  };

  return (
    <div className="d-none d-md-block mt-2">
      <div className="d-flex flex-row justify-content-start align-items-baseline">
        <div className="me-3">Filters:</div>
        <div role="list" className={`position-relative ${styles.badges}`}>
          {Object.keys(props.filters).map((category: string) => {
            return (
              <Fragment key={`filter_${category}`}>
                {props.filters[category].map((filter: string | number) => {
                  const filterName = getFilterName(category as FilterKind, filter);
                  return (
                    <span
                      role="listitem"
                      className={`badge bg-secondary rounded-0 text-light me-3 my-1 ${styles.badge} lightBorder`}
                      key={`filter_${category}_${filter}`}
                    >
                      <div className="d-flex flex-row align-items-baseline">
                        <div className={styles.content}>
                          <small className="text-uppercase fw-normal me-2">{category}:</small>
                          {filterName}
                        </div>
                        <button
                          className={`btn btn-link btn-sm lh-1 ${styles.btn}`}
                          onClick={() => props.onChange(category, filter as string, false)}
                          aria-label={`Remove ${filterName} filter`}
                        >
                          <IoMdCloseCircleOutline />
                        </button>
                      </div>
                    </span>
                  );
                })}
              </Fragment>
            );
          })}
        </div>
      </div>
    </div>
  );
};

export default SelectedFilters;
