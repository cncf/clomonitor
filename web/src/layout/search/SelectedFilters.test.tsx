import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import SelectedFilters from './SelectedFilters';

const mockOnChange = jest.fn();

const defaultProps = {
  filters: { maturity: ['2'], rating: ['a', 'b'], category: ['0', '1'] },
  onChange: mockOnChange,
};

describe('SelectedFilters', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<SelectedFilters {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      render(<SelectedFilters {...defaultProps} />);

      expect(screen.getByText('Filters:')).toBeInTheDocument();
      expect(screen.getByRole('list')).toBeInTheDocument();
      expect(screen.getAllByRole('listitem')).toHaveLength(5);

      expect(screen.getByText('maturity:')).toBeInTheDocument();
      expect(screen.getByText('Sandbox')).toBeInTheDocument();
      expect(screen.getAllByText('rating:')).toHaveLength(2);
      expect(screen.getByText('A [75-100]')).toBeInTheDocument();
      expect(screen.getByText('B [50-74]')).toBeInTheDocument();
      expect(screen.getAllByText('category:')).toHaveLength(2);
      expect(screen.getByText('App definition')).toBeInTheDocument();
      expect(screen.getByText('Observability')).toBeInTheDocument();

      expect(screen.getAllByRole('button')).toHaveLength(5);
      expect(screen.getByRole('button', { name: 'Remove Sandbox filter' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Remove A [75-100] filter' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Remove B [50-74] filter' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Remove App definition filter' })).toBeInTheDocument();
      expect(screen.getByRole('button', { name: 'Remove Observability filter' })).toBeInTheDocument();
    });

    it('calls on change', () => {
      render(<SelectedFilters {...defaultProps} />);

      const btn = screen.getByRole('button', { name: 'Remove B [50-74] filter' });
      userEvent.click(btn);

      expect(mockOnChange).toHaveBeenCalledTimes(1);
      expect(mockOnChange).toHaveBeenCalledWith('rating', 'b', false);
    });
  });

  describe('Does not render', () => {
    it('when filters is empty', () => {
      const { container } = render(<SelectedFilters {...defaultProps} filters={{}} />);
      expect(container).toBeEmptyDOMElement();
    });
  });
});
