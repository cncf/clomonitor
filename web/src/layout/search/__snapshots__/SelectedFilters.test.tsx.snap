// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectedFilters creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-none d-md-block mt-2"
  >
    <div
      class="d-flex flex-row justify-content-start align-items-baseline"
    >
      <div
        class="me-3"
      >
        Filters:
      </div>
      <div
        class="position-relative badges"
        role="list"
      >
        <span
          class="badge bg-secondary rounded-0 text-light me-3 my-1 badge lightBorder"
          role="listitem"
        >
          <div
            class="d-flex flex-row align-items-baseline"
          >
            <div
              class="content"
            >
              <small
                class="text-uppercase fw-normal me-2"
              >
                maturity:
              </small>
              Sandbox
            </div>
            <button
              aria-label="Remove Sandbox filter"
              class="btn btn-link btn-sm lh-1 btn"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 90c44.3 0 86 17.3 117.4 48.6C404.7 170 422 211.7 422 256s-17.3 86-48.6 117.4C342 404.7 300.3 422 256 422s-86-17.3-117.4-48.6C107.3 342 90 300.3 90 256s17.3-86 48.6-117.4C170 107.3 211.7 90 256 90m0-42C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
                />
                <path
                  d="M360 330.9L330.9 360 256 285.1 181.1 360 152 330.9l74.9-74.9-74.9-74.9 29.1-29.1 74.9 74.9 74.9-74.9 29.1 29.1-74.9 74.9z"
                />
              </svg>
            </button>
          </div>
        </span>
        <span
          class="badge bg-secondary rounded-0 text-light me-3 my-1 badge lightBorder"
          role="listitem"
        >
          <div
            class="d-flex flex-row align-items-baseline"
          >
            <div
              class="content"
            >
              <small
                class="text-uppercase fw-normal me-2"
              >
                rating:
              </small>
              A [75-100]
            </div>
            <button
              aria-label="Remove A [75-100] filter"
              class="btn btn-link btn-sm lh-1 btn"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 90c44.3 0 86 17.3 117.4 48.6C404.7 170 422 211.7 422 256s-17.3 86-48.6 117.4C342 404.7 300.3 422 256 422s-86-17.3-117.4-48.6C107.3 342 90 300.3 90 256s17.3-86 48.6-117.4C170 107.3 211.7 90 256 90m0-42C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
                />
                <path
                  d="M360 330.9L330.9 360 256 285.1 181.1 360 152 330.9l74.9-74.9-74.9-74.9 29.1-29.1 74.9 74.9 74.9-74.9 29.1 29.1-74.9 74.9z"
                />
              </svg>
            </button>
          </div>
        </span>
        <span
          class="badge bg-secondary rounded-0 text-light me-3 my-1 badge lightBorder"
          role="listitem"
        >
          <div
            class="d-flex flex-row align-items-baseline"
          >
            <div
              class="content"
            >
              <small
                class="text-uppercase fw-normal me-2"
              >
                rating:
              </small>
              B [50-74]
            </div>
            <button
              aria-label="Remove B [50-74] filter"
              class="btn btn-link btn-sm lh-1 btn"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 90c44.3 0 86 17.3 117.4 48.6C404.7 170 422 211.7 422 256s-17.3 86-48.6 117.4C342 404.7 300.3 422 256 422s-86-17.3-117.4-48.6C107.3 342 90 300.3 90 256s17.3-86 48.6-117.4C170 107.3 211.7 90 256 90m0-42C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
                />
                <path
                  d="M360 330.9L330.9 360 256 285.1 181.1 360 152 330.9l74.9-74.9-74.9-74.9 29.1-29.1 74.9 74.9 74.9-74.9 29.1 29.1-74.9 74.9z"
                />
              </svg>
            </button>
          </div>
        </span>
        <span
          class="badge bg-secondary rounded-0 text-light me-3 my-1 badge lightBorder"
          role="listitem"
        >
          <div
            class="d-flex flex-row align-items-baseline"
          >
            <div
              class="content"
            >
              <small
                class="text-uppercase fw-normal me-2"
              >
                category:
              </small>
              App definition
            </div>
            <button
              aria-label="Remove App definition filter"
              class="btn btn-link btn-sm lh-1 btn"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 90c44.3 0 86 17.3 117.4 48.6C404.7 170 422 211.7 422 256s-17.3 86-48.6 117.4C342 404.7 300.3 422 256 422s-86-17.3-117.4-48.6C107.3 342 90 300.3 90 256s17.3-86 48.6-117.4C170 107.3 211.7 90 256 90m0-42C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
                />
                <path
                  d="M360 330.9L330.9 360 256 285.1 181.1 360 152 330.9l74.9-74.9-74.9-74.9 29.1-29.1 74.9 74.9 74.9-74.9 29.1 29.1-74.9 74.9z"
                />
              </svg>
            </button>
          </div>
        </span>
        <span
          class="badge bg-secondary rounded-0 text-light me-3 my-1 badge lightBorder"
          role="listitem"
        >
          <div
            class="d-flex flex-row align-items-baseline"
          >
            <div
              class="content"
            >
              <small
                class="text-uppercase fw-normal me-2"
              >
                category:
              </small>
              Observability
            </div>
            <button
              aria-label="Remove Observability filter"
              class="btn btn-link btn-sm lh-1 btn"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 90c44.3 0 86 17.3 117.4 48.6C404.7 170 422 211.7 422 256s-17.3 86-48.6 117.4C342 404.7 300.3 422 256 422s-86-17.3-117.4-48.6C107.3 342 90 300.3 90 256s17.3-86 48.6-117.4C170 107.3 211.7 90 256 90m0-42C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48z"
                />
                <path
                  d="M360 330.9L330.9 360 256 285.1 181.1 360 152 330.9l74.9-74.9-74.9-74.9 29.1-29.1 74.9 74.9 74.9-74.9 29.1 29.1-74.9 74.9z"
                />
              </svg>
            </button>
          </div>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
